File: flureadium/ios/flureadium/Sources/flureadium/ReadiumReaderChannel.swift
import Flutter
import ReadiumShared

class ReadiumReaderChannel: FlutterMethodChannel {
  // Compiles fine without this init, but then mysteriously crashes later with EXC_BAD_ACCESS when calling any member function later.
  init(name: String, binaryMessenger messenger: FlutterBinaryMessenger) {
    super.init(
      name: name, binaryMessenger: messenger, codec: FlutterStandardMethodCodec.sharedInstance(),
      taskQueue: nil)
  }

  func onPageChanged(locator: Locator) {
    invokeMethod("onPageChanged", arguments: locator.jsonString as String?)
  }

  func onExternalLinkActivated(url: URL) {
    invokeMethod("onExternalLinkActivated", arguments: url.absoluteString as String?)
  }

  // Per-view delivery for selection/decoration. The global EventChannels share
  // one channel name across all reader instances, so any instance's
  // subscribe/cancel clobbers the others. This per-view MethodChannel is tied
  // to a single reader, mirroring onPageChanged, and avoids that interference.
  func onSelectionChanged(locatorJson: String?) {
    invokeMethod("onSelectionChanged", arguments: locatorJson)
  }

  func onDecorationActivated(eventJson: String?) {
    invokeMethod("onDecorationActivated", arguments: eventJson)
  }
}

File: flureadium/android/src/main/kotlin/dev/mulev/flureadium/ReadiumReaderChannel.kt
package dev.mulev.flureadium

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.util.AbsoluteUrl
import org.readium.r2.shared.util.toUri

internal class ReadiumReaderChannel(messenger: BinaryMessenger, name: String) :
    MethodChannel(messenger, name) {
    fun onPageChanged(locator: Locator?) =
        invokeMethod("onPageChanged", locator?.toJSON().toString())

    fun onExternalLinkActivated(url: AbsoluteUrl) =
        invokeMethod("onExternalLinkActivated", url.toString())

    // Per-view delivery for selection/decoration. The global EventChannels share
    // one channel name across all reader instances, so any instance's
    // subscribe/cancel clobbers the others. This per-view MethodChannel is tied
    // to a single reader, mirroring onPageChanged, and avoids that interference.
    fun onSelectionChanged(locator: Locator?) =
        invokeMethod("onSelectionChanged", locator?.toJSON().toString())

    fun onDecorationActivated(eventJson: String?) =
        invokeMethod("onDecorationActivated", eventJson)
}
